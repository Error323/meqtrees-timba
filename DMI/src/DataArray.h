//  DataArray.cc: DMI Array class (using AIPS++ Arrays)
//
//  Copyright (C) 2002
//  ASTRON (Netherlands Foundation for Research in Astronomy)
//  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//  $Id$
//
//  $Log$
//  Revision 1.17  2002/12/06 15:29:02  smirnov
//  %[BugId: 112]%
//  Fixed some bugs in the AIPS++ hooks;
//  fixed size returned from DataRecord iterator.
//
//  Revision 1.16  2002/12/05 10:15:22  smirnov
//  %[BugId: 112]%
//  Fixed Lorray support in DataArrays, etc.
//  Revised AIPS++ hooks.
//
//  Revision 1.15  2002/12/03 20:36:14  smirnov
//  %[BugId: 112]%
//  Ported DMI to use Lorrays (with blitz arrays)
//
//  Revision 1.14  2002/10/29 16:28:17  smirnov
//  %[BugId: 26]%
//  Added missing TypeIter-DMI.h file.
//  Removed spurious #include of BlockSet1.h (thanks a lot, Rose...)
//
//  Revision 1.13  2002/10/29 13:10:38  smirnov
//  %[BugId: 26]%
//  Re-worked build_aid_maps.pl and TypeIterMacros.h to enable on-demand
//  importing of data types from other packages. Basically, data types from package
//  X will be pulled in by NestableContainer only when included from
//  package Y that has an explicit dependence on package X. DMI itself depends
//  only on Common.
//
//  Migrated to new Rose C++ add-in, so all Rose markup has changed.
//
//  Revision 1.12  2002/06/11 12:15:08  smirnov
//  %[BugId: 26]%
//  Further fixes to array-mode hook addressing.
//  Added an optional tid argument to NestableContainer::size().
//
//  Revision 1.11  2002/06/10 12:39:18  smirnov
//  %[BugId: 26]%
//  Revised NestableContainer::get() interface to return info in a ContentInfo
//  structure.
//  Added optional size argument to hook.as_type_p() and _wp() methods.
//  Cleaned up size handling, added working as_vector<T> and =(vector<T>).
//
//  Revision 1.10  2002/06/07 14:22:48  smirnov
//  %[BugId: 26]%
//  Many revisions related to support of arrays and vectors (including AIPS++) by
//  hooks. Checking in now because I plan to modify the NestableContainer interface.
//
//  Revision 1.9  2002/05/30 12:15:13  diepen
//
//  %[BugId: 25]%
//  Added the required constructors
//
//  Revision 1.8  2002/05/14 09:48:00  gvd
//  Removed public which was left in
//
//  Revision 1.7  2002/05/07 13:48:52  oms
//  Minor fixes for the new build environment.
//  The code has been re-generated by Rose, hence the world+dog has
//  been updated.
//
//  Revision 1.6  2002/05/07 11:46:00  gvd
//  The 'final' version supporting array subsets
//
//  Revision 1.5  2002/04/17 12:19:31  oms
//  Added the "Intermediate" type category (for Array_xxx) and support for it
//  in hooks.
//
//  Revision 1.4  2002/04/12 10:15:09  oms
//  Added fcomplex and dcomplex types.
//  Changes to NestableContainer::get():
//   - merged autoprivatize and must_write args into a single flags arg
//   - added NC_SCALAR and NC_POINTER flags that are passed to get()
//  Got rid of isScalar() and isContiguous(), checking is now up to get().
//
//  Revision 1.3  2002/04/12 07:47:53  oms
//  Added fcomplex and dcomplex types
//
//  Revision 1.2  2002/04/08 14:27:07  oms
//  Added isScalar(tid) to DataArray.
//  Fixed isContiguous() in DataField.
//
//  Revision 1.1  2002/04/05 13:05:46  gvd
//  First version
//


#ifndef DMI_DATAARRAY_H
#define DMI_DATAARRAY_H

#include "Common/Lorrays.h"
#include "Common/Thread/Mutex.h"
#include "Common/Thread/Key.h"
#include "DMI/Common.h"
#include "DMI/DMI.h"
#include "DMI/NestableContainer.h"
#include "DMI/HIID.h"
#include "DMI/SmartBlock.h"

#ifdef HAVE_AIPSPP
#include <aips/Arrays.h>
#endif

#pragma types #DataArray


// We assume Blitz support here, AIPS++ can be re-integrated later
#ifndef LORRAYS_USE_BLITZ
  #error This version of DataArray requires Blitz Lorrays
#endif

//##ModelId=3DB949AE00C5
class DataArray : public NestableContainer
{
public:
  // Create the object without an array in it.
    //##ModelId=3DB949AE039F
  explicit DataArray (int flags = DMI::WRITE);

  // Create the object with an array of the given shape.
    //##ModelId=3DB949AE03A4
  DataArray (TypeId type, const LoShape & shape, int flags = DMI::WRITE,
	     int shm_flags = 0);

  // Create the object with an array of the given shape.
//   explicit DataArray (const Array<bool>& array, int flags = DMI::WRITE,
// 		      int shm_flags = 0);
//   explicit DataArray (const Array<int>& array, int flags = DMI::WRITE,
// 		      int shm_flags = 0);
//   explicit DataArray (const Array<float>& array, int flags = DMI::WRITE,
// 		      int shm_flags = 0);
//   explicit DataArray (const Array<double>& array, int flags = DMI::WRITE,
// 		      int shm_flags = 0);
//   explicit DataArray (const Array<fcomplex>& array, int flags = DMI::WRITE,
// 		      int shm_flags = 0);
//   explicit DataArray (const Array<dcomplex>& array, int flags = DMI::WRITE,
// 		      int shm_flags = 0);
//   
  // templated method to create a copy of the given Lorray.
  // We make use of the fact that a Lorray(N,T) is actually a blitz::Array<T,N>.
  // Hence this templated definition is equivalent to a bunch of non-templated
  // ones, each with its own type and rank.
  // For non-templated compilers, this can be redefined using the DoFor...()
  // type iterator macros
  template<class T,int N>
  explicit DataArray (const blitz::Array<T,N> & array, int flags = DMI::WRITE,
		      int shm_flags = 0);

#ifdef HAVE_AIPSPP
  // templated method to create a copy of the given AIPS++ array
  template<class T>
  explicit DataArray (const Array<T> & array, int flags = DMI::WRITE,
		      int shm_flags = 0);
#endif

  // Copy (copy semantics).
    //##ModelId=3DB949AE03AF
  DataArray (const DataArray& other, int flags = 0, int depth = 0);

    //##ModelId=3DB949AE03B8
  ~DataArray();

  // Assignment (copy semantics).
    //##ModelId=3DB949AE03B9
  DataArray& operator= (const DataArray& other);

  // True if the object contains an initialized array
  //##ModelId=3DB949AF0022
  bool valid() const;
      
  // returns rank of array
  int rank () const;
  
  // returns shape of array
  const LoShape & shape () const;
  
  // returns type of array element
  // (the virtual type() method, below, overriding the abstract one in 
  // NestableContainer, will return the array type)
  TypeId elementType () const;
  
#ifdef HAVE_AIPSPP
  // Returns contents as an AIPS++ array (by copy or reference)
  // The dummy T* argument is to help in template instantiation
  // (gcc has a problem resolving stuff like:
  //
  //    template<class T> T function_one ();
  //    template<class T> T function_two ()
  //    { return function_one<T>(); }
  //  ... the <T> in the invocation is considered an error. The dummy
  //  argument allows you to say "function_one((T*)0)" instead)
  template<class T>
  Array<T> refAipsArray (const T*);
  template<class T>
  Array<T> copyAipsArray (const T*) const;
#endif

  // Return the object type (TpDataArray).
    //##ModelId=3DB949AE03BE
  virtual TypeId objectType() const;

  // Reconstruct the DataArray object from a BlockSet.
    //##ModelId=3DB949AE03C0
  virtual int fromBlock (BlockSet& set);

  // Add the DataArray object to the BlockSet.
    //##ModelId=3DB949AE03C5
  virtual int toBlock (BlockSet& set) const;

  // Clone the object.
    //##ModelId=3DB949AE03CB
  virtual CountedRefTarget* clone (int flags = 0, int depth = 0) const;

  // Privatize the object.
    //##ModelId=3DB949AE03D2
  virtual void privatize (int flags = 0, int depth = 0);

  // Get the 
    //##ModelId=3DB949AE03DA
  virtual const void* get (const HIID& id, ContentInfo &info,
			   TypeId check_tid = 0, int flags = 0) const;

  // Insertion is not possible (throws exception).
    //##ModelId=3DB949AE03E5
  virtual void* insert (const HIID& id, TypeId tid, TypeId& real_tid);

  // The size is the number of array elements.
    //##ModelId=3DB949AF0007
  virtual int size (TypeId tid = 0) const;

  // The actual type of the array (TpArray_float, etc.).
    //##ModelId=3DB949AF000C
  virtual TypeId type() const;
  
  // Parse a HIID describing a subset and fill start,end,incr.
  // NB: "end" is one past ending position at each axis: 
  //        i.e. [start,end) is the interval, using conventional notation.
  // It fills in keepAxes telling which axes should be kept in the
  // subarray. Returns number of axes to keep.
    //##ModelId=3DB949AF000E
  int parseHIID (const HIID& id, LoPos& st, LoPos& end,LoPos& incr, 
                 vector<bool> &keepAxes) const;
  
    //##ModelId=3DB949AF001C
  DefineRefTypes(DataArray,Ref);

  static const int NumTypes = Tpbool_int - Tpstring_int + 1;
  
private:
  // Initialize internal shape and create array using the given shape.
    //##ModelId=3DB949AF0024
  void init (const LoShape & shape,int flags);

  // Create the actual Array object.
  // It is created from the array data part in the SmartBlock.
    //##ModelId=3DB949AF002B
  void makeArray();

  // Clear the object (thus remove the Array).
    //##ModelId=3DB949AF002C
  void clear();

  // Clone the object.
    //##ModelId=3DB949AF002E
  void cloneOther (const DataArray& other, int flags = 0, int depth = 0);

    //##ModelId=3DB949AE036D
  LoShape    itsShape;          // actual shape
  
  TypeId     itsType;           // array TypeId
    //##ModelId=3DB949AE0379
  TypeId     itsScaType;        // scalar data type matching the array type
    //##ModelId=3DB949AE0383
  int        itsElemSize;       // #bytes of an array element
    //##ModelId=3DB949AE0389
  int        itsSize;           // total size of array (in elements)
  int        itsDataOffset;     // array data offset in SmartBlock
    //##ModelId=3DB949AE038E
  char*      itsArrayData;      // pointer to array data in SmartBlock
    //##ModelId=3DB949AE0394
  void *     itsArray;

  //##ModelId=3DB949AE0370
  BlockRef    itsData;

  // OK, setup some circus hoops. Rank & type of DataArray is set at runtime,
  // while for blitz arrays it's compile-time. So, for every blitz operation
  // required in DataArray, we'll setup an N(ranks) x N(types) matrix of 
  // function pointers, then use rank & type to call the appropriate function.
  // This matrix is called the "method table".
  
  // Methods for the method table are naturally implemented via
  // templates. Refer to DataArray.cc.
  
  // These are the actual method tables
  typedef void * (*AllocatorWithData)(void*,const LoShape &);
  typedef void * (*AllocatorDefault)();
  typedef void (*AssignWithStride)(void*,void *,const LoShape &,const LoShape &);
  typedef void (*Destructor)(void*);
  
  static AllocatorWithData    allocatorWithData   [NumTypes][MaxLorrayRank];
  static AllocatorDefault     allocatorDefault    [NumTypes][MaxLorrayRank];
  static AssignWithStride     assignerWithStride  [NumTypes][MaxLorrayRank];
  static Destructor           destructor          [NumTypes][MaxLorrayRank];
  
  // converts a type id into a numeric offset into the table above
  static int typeIndex (TypeId tid)
  { return Tpbool_int - tid.id(); }
  // These methods do a lookup & call into each method table
  static void * allocateArrayWithData (TypeId tid,void *data,const LoShape &shape )
  {
    return (*allocatorWithData[typeIndex(tid)][shape.size()-1])(data,shape);
  }
  static void assignWithStride (TypeId tid,void *ptr,void *data,const LoShape &shape,const LoShape &stride )
  {
    (*assignerWithStride[typeIndex(tid)][shape.size()-1])(ptr,data,shape,stride);
  }
  static void * allocateArrayDefault (TypeId tid,int rank)
  {
    return (*allocatorDefault[typeIndex(tid)][rank-1])();
  }
  static void destroyArray (TypeId tid,int rank,void *ptr)
  {
    (*destructor[typeIndex(tid)][rank-1])(ptr);
  }

  // Define the subarray object (for slicing into an array)
  typedef struct { void *ptr; int rank; }  SubArray;
#ifdef USE_THREADS
  // Each thread must have its own subarray pointer. Use a map to accomplish
  // this -- I would use thread keys, but the number is way too limited.
  typedef std::map<Thread::ThrID,SubArray> SubArrayMap;
  mutable SubArrayMap itsSubArrayMap;
  
  void initSubArray () const
  {}
#else
  // non-threaded mode -- keep a single entry
  mutable SubArray itsSubArray;
  
  void initSubArray () const
  { itsSubArray.ptr = 0; }
#endif
  // this helper function creates the subarray object with the given data,
  // shape & stride. 
  void * makeSubArray (void *data,const LoShape & shape,const LoShape &stride) const;
  
#ifdef HAVE_AIPSPP
  // helper function to implement templates below without resorting to a
  // specialization (which seems to cause redefined symbol trouble)
  template<class T>
  static bool isStringArray (const Array<T> &);
  void copyStringArray (const void *source);
#endif
};

DefineRefTypes(DataArray,DataArrayRef);

inline bool DataArray::valid () const
{
  return itsArray != 0 ;
}

    
// returns rank of array
inline int DataArray::rank () const
{
  return itsShape.size();
}
  
// returns shape of array
inline const LoShape & DataArray::shape () const
{
  return itsShape;
}
  
// returns type of array element
// (the virtual type() method, below, overriding the abstract one in 
// NestableContainer, will return the array type)
inline TypeId DataArray::elementType () const
{
  return itsScaType;
}

// templated constructor from a Blitz array
template<class T,int N>
DataArray::DataArray (const blitz::Array<T,N>& array,
		      int flags, int )  // shm_flags not yet used
: NestableContainer(flags&DMI::WRITE != 0),
  itsArray    (0)
{
  initSubArray();
  itsScaType  = typeIdOf(T);
  itsElemSize = sizeof(T);
  itsType     = typeIdOfArray(T,N);
  init(array.shape(),flags);
  // after an init, itsArray contains a valid array of the given shape,
  // so we can assign the other array to it, to copy the data over
  *static_cast<blitz::Array<T,N>*>(itsArray) = array;
}

#ifdef HAVE_AIPSPP
template<class T>
inline bool DataArray::isStringArray (const Array<T> &)
{ return False; }
  
template<>
inline bool DataArray::isStringArray (const Array<String> &)
{ return True; }

inline void DataArray::copyStringArray (const void *source)
{
  const String *data = static_cast<const String*>(source);
  string *dest = reinterpret_cast<string *>(itsArrayData),*end = dest + itsSize;
  for( ; dest < end; dest++,data++ )
    *dest = *data;
}

// templated constructor from an AIPS++ array
template<class T>
DataArray::DataArray (const Array<T> &array,int flags, int )  // shm_flags not yet used
: NestableContainer(flags&DMI::WRITE != 0),
  itsArray    (0)
{
  initSubArray();
  itsScaType  = isStringArray(array) ? Tpstring : typeIdOf(T);
  itsElemSize = isStringArray(array) ? sizeof(string) : sizeof(T);
  itsType     = TpArray(itsScaType,array.ndim());
  init(LoShape(array.shape()),flags);
  // after an init, itsArray contains a valid array of the given shape,
  // so we can copy the data over
  // BUG here! use a more efficient AIPS++ array iterator
  bool del;
  const T *data = array.getStorage(del);
  if( isStringArray(array) )
    copyStringArray(data);
  else
    memcpy(itsArrayData,data,itsSize*itsElemSize);
  array.freeStorage(data,del);
}


template<class T>
Array<T> DataArray::copyAipsArray (const T*) const
{
  FailWhen( !valid(),"invalid DataArray" );
  if( isStringArray(Array<T>()) )
  {
    FailWhen( itsScaType != Tpstring,"array type mismatch" );
    String *dest = new String[itsSize], *end = dest+itsSize;
    const string *src = reinterpret_cast<const string *>(itsArrayData);
    for( ; dest < end; dest++,src++ )
      *dest = *src;
    // The reinterpret_cast is necessary to prevent confusing the compiler.
    // In reality this code gets executed only when T=String
    return Array<T>(itsShape,reinterpret_cast<T*>(dest),TAKE_OVER);
  }
  else
  {
    FailWhen( itsScaType != typeIdOf(T),"array type mismatch" );
    return Array<T>(itsShape,reinterpret_cast<const T*>(itsArrayData));
  }
}

template<class T>
Array<T> DataArray::refAipsArray (const T*)
{
  FailWhen( !valid(),"invalid DataArray" );
  FailWhen( !isWritable(),"r/w access violation" );
  if( isStringArray(Array<T>()) )
    return copyAipsArray((T*)0);
  FailWhen( itsScaType != typeIdOf(T),"array type mismatch" );
  return Array<T>(itsShape,itsArrayData,SHARE);
}

#endif

#endif
