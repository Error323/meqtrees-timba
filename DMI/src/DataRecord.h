//## begin module%1.4%.codegen_version preserve=yes
//   Read the documentation to learn more about C++ code generator
//   versioning.
//## end module%1.4%.codegen_version

//## begin module%3C10CC820052.cm preserve=no
//	  %X% %Q% %Z% %W%
//## end module%3C10CC820052.cm

//## begin module%3C10CC820052.cp preserve=no
//## end module%3C10CC820052.cp

//## Module: DataRecord%3C10CC820052; Package specification
//## Subsystem: DMI%3C10CC810155
//	f:\lofar\dvl\lofar\cep\cpa\pscf\src
//## Source file: F:\LOFAR\dvl\LOFAR\cep\cpa\pscf\src\DataRecord.h

#ifndef DataRecord_h
#define DataRecord_h 1

//## begin module%3C10CC820052.additionalIncludes preserve=no
#include "Common.h"
#include "DMI.h"
//## end module%3C10CC820052.additionalIncludes

//## begin module%3C10CC820052.includes preserve=yes
//## end module%3C10CC820052.includes

// NestableContainer
#include "NestableContainer.h"
// HIID
#include "HIID.h"
// DataField
#include "DataField.h"
//## begin module%3C10CC820052.declarations preserve=no
//## end module%3C10CC820052.declarations

//## begin module%3C10CC820052.additionalDeclarations preserve=yes
//## end module%3C10CC820052.additionalDeclarations


//## begin DataRecord%3BB3112B0027.preface preserve=yes
//## end DataRecord%3BB3112B0027.preface

//## Class: DataRecord%3BB3112B0027
//## Category: PSCF::DMI%3BEAB1F2006B; Global
//## Subsystem: DMI%3C10CC810155
//## Persistence: Transient
//## Cardinality/Multiplicity: n



//## Uses: <unnamed>%3BFCD23E02FB;HIID { -> }

class DataRecord : public NestableContainer  //## Inherits: <unnamed>%3BFCD87E039E
{
  //## begin DataRecord%3BB3112B0027.initialDeclarations preserve=yes
  //## end DataRecord%3BB3112B0027.initialDeclarations

  public:
    //## Constructors (generated)
      DataRecord();

      DataRecord(const DataRecord &right);

    //## Destructor (generated)
      ~DataRecord();

    //## Assignment Operation (generated)
      DataRecord & operator=(const DataRecord &right);


    //## Other Operations (specified)
      //## Operation: field%3BFBF49D00A1
      DataFieldRef & field (const HIID &id);

      //## Operation: operator []%3BFBF55C02A4
      DataFieldRef & operator [] (const HIID &id);

      //## Operation: add%3BFBF5B600EB
      Bool add (const HIID &id, DataFieldRef &ref);

      //## Operation: remove%3BB311C903BE
      Bool remove (const HIID &id);

      //## Operation: replace%3BFCD4BB036F
      Bool replace (const HIID &id, DataFieldRef &ref);

      //## Operation: enableWrite%3BFCF8C00234
      Bool enableWrite (const HIID &id);

    //## Get and Set Operations for Class Attributes (generated)

      //## Attribute: writable%3BFCD97F01DA
      const Bool isWritable () const;

    // Additional Public Declarations
      //## begin DataRecord%3BB3112B0027.public preserve=yes
      //## end DataRecord%3BB3112B0027.public

  protected:
    // Additional Protected Declarations
      //## begin DataRecord%3BB3112B0027.protected preserve=yes
      //## end DataRecord%3BB3112B0027.protected

  private:
    // Additional Private Declarations
      //## begin DataRecord%3BB3112B0027.private preserve=yes
      //## end DataRecord%3BB3112B0027.private

  private: //## implementation
    // Data Members for Class Attributes

      //## begin DataRecord::writable%3BFCD97F01DA.attr preserve=no  public: Bool {U} 
      Bool writable;
      //## end DataRecord::writable%3BFCD97F01DA.attr

    // Data Members for Associations

      //## Association: PSCF::<unnamed>%3BE123050350
      //## Role: DataRecord::fields%3BE123060149
      //## begin DataRecord::fields%3BE123060149.role preserve=no  private: DataField { -> 0..nVHgN}
      map<HIID,DataFieldRef> fields;
      //## end DataRecord::fields%3BE123060149.role

    // Additional Implementation Declarations
      //## begin DataRecord%3BB3112B0027.implementation preserve=yes
      //## end DataRecord%3BB3112B0027.implementation

};

//## begin DataRecord%3BB3112B0027.postscript preserve=yes
//## end DataRecord%3BB3112B0027.postscript

// Class DataRecord 

//## Get and Set Operations for Class Attributes (inline)

inline const Bool DataRecord::isWritable () const
{
  //## begin DataRecord::isWritable%3BFCD97F01DA.get preserve=no
  return writable;
  //## end DataRecord::isWritable%3BFCD97F01DA.get
}

//## begin module%3C10CC820052.epilog preserve=yes
//## end module%3C10CC820052.epilog


#endif
