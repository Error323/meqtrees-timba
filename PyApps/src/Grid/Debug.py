from Timba import utils

_dbg = utils.verbosity(0,name='gw');
_dprint = _dbg.dprint;
_dprintf = _dbg.dprintf;

__all__ = [ '_dbg','_dprint','_dprintf' ];
