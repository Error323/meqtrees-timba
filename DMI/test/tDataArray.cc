#include "DataArray.h"
#include "DataRecord.h"
#include "Debug.h"
#include <aips/Arrays/ArrayMath.h>

int main()
{
  try {
    DataRecord rec;
    rec["A"] <<= new DataArray(TpArray_float, IPosition(2,10,12));
    float* ptr1 = rec["A/5.5"].as_float_wp();
    Array_float* farr = &rec["A"];
    Assert (farr->nelements() == 20*12);
    indgen (*farr);
    float* ptr = &rec["A"];
    for (int i=0; i<10*12; i++) {
      Assert (ptr[i] == i);
    }
  } catch (Debug::Error& x) {
    cerr << x.what() << endl;
    return 1;
  }
  return 0;
}
