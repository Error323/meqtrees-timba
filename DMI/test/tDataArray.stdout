
Expect type mismatch exception ...
>>>
[DataArray/082a4d38 RW Array(float,2)  10x12] array type mismatch (Array(double,2)/Array(float,2)) (context: Global)
<<<

Expect type mismatch exception ...
>>>
[DataArray/082a4d38 RW Array(float,2)  10x12] array type mismatch (double/Array(float,2)) (context: Global)
<<<
2[ 0 0 ][ 10 12 ][ 1 1 ][ 1 1 ]
0[ 1 2 ][ 2 3 ][ 1 1 ][ 0 0 ]
1[ 1 0 ][ 2 12 ][ 1 1 ][ 0 1 ]
1[ 0 2 ][ 10 3 ][ 1 1 ][ 1 0 ]
2[ 1 0 ][ 4 12 ][ 1 2 ][ 1 1 ]

Expect some HIID parser exceptions ...
Only 3 values (start:end:incr) can be given per axis in HIID 1:3._:_:_:2
Start 10 should be < axis length 10 in HIID 10._
Start 12 should be < axis length 12 in HIID _.12
End 9 should be >= start 10 in HIID _.10:9
Stride 0 should be > 0 in HIID _.9:10:0
HIID _._._ has more axes than the array (2)
HIID : does not have a value before range delimiter
HIID 1:: does not have a value before range delimiter
HIID 1: does not have a value after range delimiter
